src/porepy/numerics/ad/secondary_operator.py

```diff
@@ -0,0 +1,897 @@
+
+        """
+        vals = []
+        for _, data in self.mdg.interfaces(return_data=True):
+            vals.append(pp.get_solution_values(self._name_diffs, data, iterate_index=0))
+        if len(vals) > 0:
+            return np.hstack(vals)
+        else:
+            return np.zeros(self.num_dependencies, dtype=float)
+
+    @interface_derivatives.setter
+    def interface_derivatives(self, val: np.ndarray) -> None:
+
+        i = 0
+        for grid in self.mdg.interfaces():
+            n = grid.num_cells
+            self.progress_iterate_derivatives_on_grid(val[:, i : i + n], grid)
+            i += n
+
+    # Methods to progress values and derivatives in time on subdomains or
+    # interfaces (don't need new values)
+
+    def progress_values_in_time(
+        self, domains: Sequence[pp.Grid | pp.MortarGrid]
+    ) -> None:
+        """Shifts timestepping values backwards in times and sets the current
+        iterate value as the (first) previous time step value.
+
+        Does this only if the expression was instatiated as time-dependent.
+
+        Parameters:
+            domains: Performs the progress on given sequence of domains.
+
+        """
+        if self._time_dependent:
+            for grid in domains:
+                data = self._data_of(grid)
+                pp.shift_solution_values(self.name, data, pp.TIME_STEP_SOLUTIONS)
+                current_vals = pp.get_solution_values(self.name, data, iterate_index=0)
+                pp.set_solution_values(self.name, current_vals, data, time_step_index=1)
+
+    def progress_derivatives_in_time(
+        self, domains: Sequence[pp.Grid | pp.MortarGrid]
+    ) -> None:
+        """Analogous to :meth:`progress_values_in_time`, but for derivatives."""
+        if self._time_dependent:
+            for grid in domains:
+                data = self._data_of(grid)
+                pp.shift_solution_values(self._name_diffs, data, pp.TIME_STEP_SOLUTIONS)
+                current_vals = pp.get_solution_values(
+                    self._name_diffs, data, iterate_index=0
+                )
+                pp.set_solution_values(
+                    self._name_diffs, current_vals, data, time_step_index=1
+                )
+
+    # Methods operating on single grids
+
+    def update_boundary_values(
+        self, values: np.ndarray, boundary_grid: pp.BoundaryGrid
+    ) -> None:
+        """Function to update the value of the secondary expression on the boundary.
+
+        The update process of boundary values is different from the process on
+        subdomains and interfaces.
+
+        1. Boundary values have only a single iterate value (current time step)
+        2. Boundary values can have multiple previous time steps, if expression is
+           time-dependent.
+
+        Therefore, values for the current time can always be set, but values will be
+        shifted backwards in time only if the quantity is time-dependent.
```

Comment:
How far back will it be shifted? From my reading of `shift_solution_values`, it will keep the length of the list (or dictionary). Does this mean that the user is expected to initialize the list of boundary values with the correct number of items? If so, this should be documented, should it not? The same applies (I believe) to subdomain and interfaces, both iterates and time steps.

Let me know if you want to discuss this.

Reply:
`shift_solution_values` will extend the length by 1, because `i -> i + 1`.
Then the entry at `i= 0` and `i = 1` are identical, until the user provides a new value (both time and iter sense).

This is related to the discussion with @IvarStefansson  on the indexation. 
We probably should update that first in `develop`, before adapting this class

Reply:
Agreed, let's do the cleanup before polishing this PR.